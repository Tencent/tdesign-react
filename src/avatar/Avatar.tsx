import React, { useRef, useState, useEffect, useContext, Ref } from 'react';
import useResizeObserver from 'use-resize-observer';
import classNames from 'classnames';
import useConfig from '../hooks/useConfig';
import forwardRefWithStatics from '../_util/forwardRefWithStatics';
import useCommonClassName from '../_util/useCommonClassName';
import composeRefs from '../_util/composeRefs';
import { TdAvatarProps } from './type';
import { StyledProps } from '../common';
import AvatarContext from './AvatarContext';
import AvatarGroup from './AvararGroup';
import { avatarDefaultProps } from './defaultProps';

export interface AvatarProps extends TdAvatarProps, StyledProps {
  children?: React.ReactNode;
}

const Avatar = forwardRefWithStatics(
  (props: AvatarProps, ref: Ref<HTMLElement>) => {
    const {
      alt,
      hideOnLoadFailed,
      icon,
      image,
      shape,
      size: avatarSize,
      onError,
      children,
      style,
      className,
      ...avatarProps
    } = props;
    const groupSize = useContext(AvatarContext);
    const { classPrefix } = useConfig();
    const [scale, setScale] = useState(1);
    const [isImgExist, setIsImgExist] = useState(true);
    const avatarRef = useRef<HTMLElement>(null);
    const avatarChildrenRef = useRef<HTMLElement>(null);
    const size = avatarSize === 'default' ? groupSize : avatarSize;
    const gap = 4;
    const handleScale = () => {
      if (!avatarChildrenRef.current || !avatarRef.current) {
        return;
      }
      const childrenWidth = avatarChildrenRef.current.offsetWidth;
      const avatarWidth = avatarRef.current.offsetWidth;

      if (childrenWidth !== 0 && avatarWidth !== 0) {
        if (gap * 2 < avatarWidth) {
          setScale(avatarWidth - gap * 2 < childrenWidth ? (avatarWidth - gap * 2) / childrenWidth : 1);
        }
      }
    };
    const { ref: observerRef } = useResizeObserver<HTMLDivElement>({
      onResize: handleScale,
    });

    const handleImgLoadError = () => {
      onError && onError();
      !hideOnLoadFailed && setIsImgExist(false);
    };

    useEffect(() => {
      setIsImgExist(true);
      setScale(1);
    }, [props.image]);

    useEffect(() => {
      handleScale();
    }, []);

    const { SIZE } = useCommonClassName();
    const numSizeStyle: React.CSSProperties =
      size && !SIZE[size]
        ? {
            width: size,
            height: size,
            fontSize: `${Number.parseInt(size, 10) / 2}px`,
          }
        : {};
    const imageStyle: React.CSSProperties =
      size && !SIZE[size]
        ? {
            width: size,
            height: size,
          }
        : {};

    const preClass = `${classPrefix}-avatar`;

    const avatarClass = classNames(preClass, className, {
      [SIZE[size]]: !!SIZE[size],
      [`${preClass}--${shape}`]: !!shape,
      [`${preClass}-icon`]: !!icon,
    });

    let content;
    if (image && isImgExist) {
      content = <img src={image} alt={alt} style={imageStyle} onError={handleImgLoadError} />;
    } else if (icon) {
      content = icon;
    } else {
      const childrenStyle: React.CSSProperties = {
        transform: `scale(${scale})`,
      };
      content = (
        <span ref={composeRefs(ref, avatarChildrenRef, observerRef) as any} style={childrenStyle}>
          {children}
        </span>
      );
    }
    return (
      <div
        ref={composeRefs(ref, avatarRef) as any}
        className={avatarClass}
        style={{ ...numSizeStyle, ...style }}
        {...avatarProps}
      >
        {content}
      </div>
    );
  },
  { Group: AvatarGroup },
);

Avatar.displayName = 'Avatar';
Avatar.defaultProps = avatarDefaultProps;

export default Avatar;
