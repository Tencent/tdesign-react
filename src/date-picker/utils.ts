import chunk from 'lodash/chunk';
import { DateValue } from './type';
import { Styles } from '../common';

/**
 * next Month
 * @since version V1.0.0
 * @param { Date } date
 * @returns {Date}
 */
export const nextMonth = (date: Date) => {
  const isPassYear = date.getMonth() === 11;
  const newMonth = isPassYear ? 0 : date.getMonth() + 1;
  const year = isPassYear ? date.getFullYear() + 1 : date.getFullYear();
  return new Date(year, newMonth, 1);
};

/**
 * prev Month
 * @since version V1.0.0
 * @param {Date} date
 * @returns {Date}
 */
export const prevMonth = (date: Date) => {
  const passYear = date.getMonth() === 0;
  const newMonth = passYear ? 11 : date.getMonth() - 1;
  const year = passYear ? date.getFullYear() - 1 : date.getFullYear();
  return new Date(year, newMonth, 1);
};

/**
 * dateIndexOf
 * @since version V1.0.0
 * @param {Date} date
 * @param {Date[]} arrDates
 * @returns {Number}
 */
export function dateIndexOf(date: Date, arrDates: DateValue[]): number {
  const curDate = new Date(date);
  const curTime = curDate.setHours(0, 0, 0, 0);
  const selectedDates = arrDates.map((d: Date) => {
    const dt = new Date(d);
    const dtime = dt.setHours(0, 0, 0, 0);
    return dtime;
  });

  // console.log('curDate', selectedDates, curDate, selectedDates.indexOf(curTime));

  return selectedDates.indexOf(curTime);
}

/**
 * Empty value
 * @since version V1.0.0
 * @param mixedVar
 * @returns {Boolean}
 * @example
 * // example 1: empty(null)
 * // returns 1: true
 * // example 2: empty(undefined)
 * // returns 2: true
 * // example 3: empty([])
 * // returns 3: true
 * // example 4: empty({})
 * // returns 4: true
 * // example 5: empty({'fn' : function () {} })
 * // returns 5: false
 */
export function empty(mixedVar: any): mixedVar is undefined | null | false | '' | {} {
  let undef: undefined;
  const emptyValues = [undef, null, false, ''];
  for (let i = 0, len = emptyValues.length; i < len; i++) {
    if (mixedVar === emptyValues[i]) {
      return true;
    }
  }

  if (typeof mixedVar === 'object') {
    return Object.keys(mixedVar).length < 1;
  }

  return false;
}

/**
 * To Dashed from Camel Case
 * @since version V1.0.0
 * @param {String} strCamelCase
 * @returns {string}
 * @example
 * toDash('strCamelCase') === 'str-camel-case'
 */
export function toDash(strCamelCase: string): string {
  return String(strCamelCase)
    .split(/(?=[A-Z])/)
    .join('-')
    .toLowerCase();
}

export function setStyles(el: HTMLElement | SVGElement, styles: Styles) {
  Object.entries(styles).forEach(([prop, val = null]) => {
    if (empty(val)) {
      el.style.removeProperty(prop);
      return;
    }
    const [value, pri = ''] = String(val).split('!');

    el.style.setProperty(toDash(prop), value, pri);
  });
}

/**
 * 首字母大写
 * @param {String} str 目标字符串
 * @returns {String}
 */
export function firstUpperCase(str: string): string {
  if (!str) return str;
  return str[0].toUpperCase().concat(str.substring(1, str.length));
}

interface DateObj {
  year: number;
  month: number;
}

/**
 * 返回指定年、月的第一天日期
 * @param {Object} { year, month }
 * @returns {Date}
 */
function getFirstDayOfMonth({ year, month }: DateObj): Date {
  return new Date(year, month, 1);
}

/**
 * 返回指定年、月的天数
 * @param {Object} { year, month }
 * @returns {Number}
 */
function getDaysInMonth({ year, month }: DateObj): number {
  return new Date(year, month + 1, 0).getDate();
}

/**
 * 返回指定年、月的最后一天日期
 * @param {Object} { year, month }
 * @returns {Date}
 */
function getLastDayOfMonth({ year, month }: DateObj): Date {
  return new Date(year, month, getDaysInMonth({ year, month }));
}

function isSameYear(date1: Date, date2: Date) {
  return date1.getFullYear() === date2.getFullYear();
}

function isSameMonth(date1: Date, date2: Date) {
  return isSameYear(date1, date2) && date1.getMonth() === date2.getMonth();
}

function isSameDate(date1: Date, date2: Date) {
  return isSameMonth(date1, date2) && date1.getDate() === date2.getDate();
}

/**
 * 是否是某法范围内的日期，精确到日
 * @param {Date} value 目标日期
 * @param {Object} { start, end } 范围
 * @returns {Boolean}
 */
function isBetween(
  value: { getFullYear: () => number; getMonth: () => number; getDate: () => number },
  { start, end }: { start: any; end: any },
): boolean {
  const date = new Date(value.getFullYear(), value.getMonth(), value.getDate());

  const startTime = new Date(start.getFullYear(), start.getMonth(), start.getDate());
  const endTime = new Date(end.getFullYear(), end.getMonth(), end.getDate());

  return startTime <= date && endTime >= date;
}

/**
 * 比较两个日期对象的时间戳
 * @param {Date} date1 日期1
 * @param {Date} date2 日期2
 * @returns {Number} 返回 date1.getTime() - date2.getTime() 的差值
 */
function compareAsc(date1: { getTime: () => any }, date2: Date): number {
  const d1 = date1.getTime();
  const d2 = date2.getTime();

  if (d1 < d2) return -1;
  if (d1 > d2) return 1;
  return 0;
}

/**
 * 比较两个 Date 是否是同一天 或则 同一月 或则 同一年
 * @param {Date} date1 比较的日期
 * @param {Date} date2 比较的日期
 * @param {String} type 比较类型，默认比较到『日』 date|month|year
 * @returns {Boolean}
 */
export function isSame(date1: Date, date2: Date, type = 'date'): boolean {
  const func = {
    isSameYear,
    isSameMonth,
    isSameDate,
  };
  return func[`isSame${firstUpperCase(type)}`](date1, date2);
}

export function outOfRanges(d: Date, min: any, max: any) {
  return (min && compareAsc(d, min) === -1) || (max && compareAsc(d, max) === 1);
}

/**
 * @returns {Date} 当天零点的日期对象
 */
export function getToday(): Date {
  const now = new Date();
  return new Date(now.getFullYear(), now.getMonth(), now.getDate(), 0, 0, 0);
}

/**
 * 返回日期对象的年、月、日、小时、分钟、秒、12小时制标识
 * @param {Date} date
 * @returns {Object}
 */
export function getDateObj(date: Date) {
  if (!(date instanceof Date)) {
    throw new Error('date must be Date object');
  }
  return {
    year: date.getFullYear(),
    month: date.getMonth(),
    date: date.getDate(),
    hours: date.getHours(),
    minutes: date.getMinutes(),
    seconds: date.getSeconds(),
    meridiem: date.getHours() > 11 ? 'PM' : 'AM',
  };
}

/**
 * 设置日期对象的时间部分
 * @param {Date} d 日期
 * @param {Number} hour 小时
 * @param {Number} min 分钟
 * @param {Number} sec 秒
 * @returns {Date} 一个新的date
 */
export function setDateTime(d: Date, hour: number, min: number, sec: number): Date {
  // eslint-disable-next-line
  const { year, month, date } = getDateObj(d);
  return new Date(year, month, date, hour, min, sec, 0);
}

/**
 * 增加月份
 * @param {Date} date 起始日期
 * @param {Number} num 月份数
 * @returns {Date}
 */
export function subtractMonth(date: Date, num: any): Date {
  const day = date.getDate();
  const newDate = new Date(date);

  let monthNum = num;
  // eslint-disable-next-line no-plusplus
  while (monthNum--) {
    newDate.setDate(0);
  }
  newDate.setDate(day);
  return newDate;
}

/**
 * 减月份
 * @param {Date} date 起始日期
 * @param {Number} num 月份数
 * @returns {Date}
 */
export function addMonth(date: Date, num: number): Date {
  let monthNum = num;
  if (num < 0) monthNum = 0;
  const newDate = new Date(date);
  newDate.setMonth(date.getMonth() + monthNum);
  return newDate;
}

export interface OptionsType {
  firstDayOfWeek: number;
  disableDate: Function;
  minDate: Date;
  maxDate: Date;
  monthText?: string;
}

export function getWeeks(
  { year, month }: { year: number; month: number },
  { firstDayOfWeek, disableDate = () => false, minDate, maxDate }: OptionsType,
) {
  const prependDay = getFirstDayOfMonth({ year, month });

  const appendDay = getLastDayOfMonth({ year, month });

  const maxDays = getDaysInMonth({ year, month });

  const daysArr = [];

  const today = getToday();

  for (let i = 1; i <= maxDays; i++) {
    const currentDay = new Date(year, month, i);

    daysArr.push({
      text: i,
      active: false,
      value: currentDay,
      disabled: disableDate(currentDay) || outOfRanges(currentDay, minDate, maxDate),
      now: isSame(today, currentDay),
      firstDayOfMonth: i === 1,
      lastDayOfMonth: i === maxDays,
      type: 'current-month',
    });
  }

  if (prependDay.getDay() !== firstDayOfWeek - 1) {
    prependDay.setDate(0); // 上一月

    const FLAG = true;
    while (FLAG) {
      daysArr.unshift({
        text: prependDay.getDate().toString(),
        active: false,
        value: new Date(prependDay),
        disabled: disableDate(prependDay) || outOfRanges(prependDay, minDate, maxDate),
        additional: true, // 非当前月
        type: 'prev-month',
      });
      prependDay.setDate(prependDay.getDate() - 1);
      if (prependDay.getDay() === Math.abs(firstDayOfWeek + 6) % 7) break;
    }
  }

  const LEN = 42; // 显示6周

  while (daysArr.length < LEN) {
    appendDay.setDate(appendDay.getDate() + 1);
    daysArr.push({
      text: appendDay.getDate(),
      active: false,
      value: new Date(appendDay),
      disabled: disableDate(appendDay) || outOfRanges(appendDay, minDate, maxDate),
      additional: true, // 非当前月
      type: 'next-month',
    });
  }

  return chunk(daysArr, 7);
}

export function getYears(year: number, { disableDate = () => false, minDate, maxDate }: OptionsType) {
  const startYear = parseInt((year / 10).toString(), 10) * 10;
  const endYear = startYear + 9;

  const yearArr = [];

  const today = getToday();

  for (let i = startYear; i <= endYear; i++) {
    const date = new Date(i, 1);

    let disabledMonth = 0;
    let outOfRangeMonth = 0;

    for (let j = 0; j < 12; j++) {
      const d = new Date(i, j);
      if (disableDate(d)) disabledMonth += 1;
      if (outOfRanges(d, minDate, maxDate)) outOfRangeMonth += 1;
    }

    yearArr.push({
      value: date,
      now: isSame(date, today, 'year'),
      disabled: disabledMonth === 12 || outOfRangeMonth === 12,
      active: false,
      text: `${date.getFullYear()}`,
    });
  }

  return chunk(yearArr, 4);
}

export function getMonths(year: number, { disableDate = () => false, minDate, maxDate, monthText = '' }: OptionsType) {
  const monthArr = [];

  const today = getToday();

  for (let i = 0; i <= 11; i++) {
    const date = new Date(year, i);

    let disabledDay = 0;
    let outOfRangeDay = 0;

    const daysInMonth = getDaysInMonth({ year, month: i });

    for (let j = 1; j <= daysInMonth; j++) {
      const d = new Date(year, i, j);
      if (disableDate(d)) disabledDay += 1;
      if (outOfRanges(d, minDate, maxDate)) outOfRangeDay += 1;
    }

    monthArr.push({
      value: date,
      now: isSame(date, today, 'month'),
      disabled: disabledDay === daysInMonth || outOfRangeDay === daysInMonth,
      active: false,
      text: `${date.getMonth() + 1} ${monthText}`,
    });
  }

  return chunk(monthArr, 4);
}

interface DateTime {
  active: boolean;
  highlight: boolean;
  startOfRange: boolean;
  endOfRange: boolean;
  value: Date;
}

export function flagActive(data: any[], { ...args }: any) {
  const { start, end, type = 'date' } = args;

  if (!end) {
    return data.map((row: any[]) =>
      row.map((item: DateTime) => {
        Object.assign(item, { active: isSame(item.value, start, type) });
        return item;
      }),
    );
  }

  return data.map((row: any[]) =>
    row.map((item: DateTime) => {
      const date = item.value;
      const isStart = isSame(start, date, type);
      const isEnd = isSame(end, date, type);
      Object.assign(item, {
        active: isStart || isEnd,
        highlight: isBetween(date, { start, end }),
        startOfRange: isStart,
        endOfRange: isEnd,
      });
      return item;
    }),
  );
}
