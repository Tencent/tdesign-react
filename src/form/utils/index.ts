import type { NamePath } from '../type';

// 获取 formMap 管理的数据
export function getMapValue(name: NamePath, formMapRef: React.MutableRefObject<Map<any, any>>) {
  if (!formMapRef.current) return;

  // 提取所有 map key
  const mapKeys = [...formMapRef.current.keys()];
  // 转译为字符串后比对 key 兼容数组格式
  const key = mapKeys.find((key) => String(key) === String(name));
  // 拿到 key 引用地址获取 value
  return formMapRef.current.get(key);
}

// 将数据整理成对象，数组 name 转化嵌套对象: ['user', 'name'] => { user: { name: '' } }
export function calcFieldValue(name: NamePath, value: any) {
  if (Array.isArray(name)) {
    const fieldValue = name.reduceRight((prev, curr) => {
      const arr = [];
      if (/^\d+$/.test(String(curr))) arr[curr] = prev;
      return arr.length ? arr : { [curr]: prev };
    }, value);
    return { ...fieldValue };
  }

  return { [name]: value };
}

// 通过对象数据类型获取 map 引用: { user: { name: '' } } => formMap.get(['user', 'name'])
export function travelMapFromObject(
  obj: Record<any, any>,
  formMapRef: React.MutableRefObject<Map<any, any>>,
  callback: Function,
) {
  for (const [mapName, formItemRef] of formMapRef.current.entries()) {
    // 支持嵌套数据结构
    if (Array.isArray(mapName)) {
      // 创建唯一临时变量 symbol
      const symbol = Symbol('name');
      let fieldValue = null;

      for (let i = 0; i < mapName.length; i++) {
        const item = mapName[i];
        if (Reflect.has(fieldValue || obj, item)) {
          fieldValue = Reflect.get(fieldValue || obj, item);
        } else {
          // 当反射无法获取到值则重置为 symbol
          fieldValue = symbol;
          break;
        }
      }

      // 非 symbol 说明获取到了值
      if (fieldValue !== symbol) {
        callback(formItemRef, fieldValue);
      }
    } else if (Reflect.has(obj, mapName)) {
      callback(formItemRef, obj[mapName]);
    }
  }
}
