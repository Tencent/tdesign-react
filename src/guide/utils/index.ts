import { scrollToParentVisibleArea } from './getScrollParent';
import getRelativePosition from './getRelativePosition';
import getTargetElm from './getTargetElm';
import scrollToElm from './scrollToElm';
import useMount from './useMount';

export { scrollToParentVisibleArea, getRelativePosition, getTargetElm, scrollToElm, useMount };
