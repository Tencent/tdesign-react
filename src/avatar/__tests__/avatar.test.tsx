import React from 'react';
import { vi, render, fireEvent } from '@test/utils';
import Avatar from '../Avatar';

describe('Avatar 组件测试', () => {
  test('Avatar 默认 DOM 结构', async () => {
    const imageSrc = 'https://tdesign.gtimg.com/site/avatar.jpg';
    const wrapper = render(<Avatar image={imageSrc} alt="test-avatar"></Avatar>);
    const image: any = wrapper.getByAltText('test-avatar');
    expect(image.src).toBe(imageSrc);
  });

  test('Avatar 初始化 Size', async () => {
    const imageSrc = 'https://tdesign.gtimg.com/site/avatar.jpg';
    const wrapper = render(<Avatar image={imageSrc} alt="test-avatar" size={'50px'}></Avatar>);
    const image = wrapper.container.querySelector('.t-image__wrapper');
    expect(image).toHaveStyle('width: 50px');
  });

  test('Avatar 初始化 Shape', async () => {
    const { container } = render(<Avatar shape="round" size="28px"></Avatar>);
    expect(container.firstChild).toHaveClass('t-avatar--round');
  });

  test('Avatar 初始化 Dom', async () => {
    const { container } = render(<Avatar shape="round" size="28px"></Avatar>);
    expect(container.firstChild).toBeInTheDocument();
  });

  test('Avatar onError 回调', async () => {
    const mockOnErrorFn = vi.fn();
    const wrapper = render(<Avatar image="http://error/" alt="test-avatar" onError={mockOnErrorFn}></Avatar>);
    const image = wrapper.getByAltText('test-avatar');
    fireEvent(image, new Event('error'));
    expect(mockOnErrorFn).toHaveBeenCalledTimes(1);
  });
});
