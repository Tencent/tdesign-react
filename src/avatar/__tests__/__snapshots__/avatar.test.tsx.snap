// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adjust.jsx 1`] = `
<DocumentFragment>
  <div
    class="demo-avatar"
  >
    <div
      class="t-avatar t-avatar--circle"
      style="margin-right: 40px;"
    >
      <span
        style="transform: scale(1);"
      >
        王
      </span>
    </div>
    <div
      class="t-avatar t-avatar--circle"
      style="margin-right: 40px;"
    >
      <span
        style="transform: scale(1);"
      >
        王亿
      </span>
    </div>
    <div
      class="t-avatar t-avatar--circle"
      style="margin-right: 40px;"
    >
      <span
        style="transform: scale(1);"
      >
        王亿亿
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`base.jsx 1`] = `
<DocumentFragment>
  <div
    class="demo-avatar"
  >
    <div
      class="t-avatar t-avatar--circle"
      style="margin-right: 40px;"
    >
      <img
        src="https://tdesign.gtimg.com/site/avatar.jpg"
      />
    </div>
    <div
      class="t-avatar t-avatar--circle"
      style="margin-right: 40px;"
    >
      <span
        style="transform: scale(1);"
      >
        W
      </span>
    </div>
    <div
      class="t-avatar t-avatar--circle t-avatar-icon"
      style="margin-right: 40px;"
    >
      <svg
        class="t-icon t-icon-user"
        fill="none"
        height="1em"
        viewBox="0 0 16 16"
        width="1em"
      >
        <path
          d="M11.5 5a3.5 3.5 0 11-7 0 3.5 3.5 0 017 0zm-1 0a2.5 2.5 0 10-5 0 2.5 2.5 0 005 0zM13.96 10.85c.34.16.54.5.54.87V14a.5.5 0 01-.5.5H2a.5.5 0 01-.5-.5v-2.28c0-.37.2-.7.54-.87a13.79 13.79 0 0111.92 0zM8 10.5c-1.97 0-3.83.45-5.5 1.24v1.76h11v-1.76A12.78 12.78 0 008 10.5z"
          fill="currentColor"
          fill-opacity="0.9"
        />
      </svg>
    </div>
    <div
      class="t-avatar t-avatar--circle"
      style="margin-right: 40px;"
    >
      <span
        style="transform: scale(1);"
      >
        <svg
          class="t-icon t-icon-user"
          fill="none"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
        >
          <path
            d="M11.5 5a3.5 3.5 0 11-7 0 3.5 3.5 0 017 0zm-1 0a2.5 2.5 0 10-5 0 2.5 2.5 0 005 0zM13.96 10.85c.34.16.54.5.54.87V14a.5.5 0 01-.5.5H2a.5.5 0 01-.5-.5v-2.28c0-.37.2-.7.54-.87a13.79 13.79 0 0111.92 0zM8 10.5c-1.97 0-3.83.45-5.5 1.24v1.76h11v-1.76A12.78 12.78 0 008 10.5z"
            fill="currentColor"
            fill-opacity="0.9"
          />
        </svg>
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`group.jsx 1`] = `
<DocumentFragment>
  <div
    class="demo-avatar"
  >
    <div
      class="demo-avatar-block"
      style="margin-bottom: 40px;"
    >
      <div
        class="t-avatar-group t-avatar--offset-right"
      >
        <div
          class="t-avatar t-avatar--circle"
        >
          <img
            src="https://tdesign.gtimg.com/site/avatar.jpg"
          />
        </div>
        <div
          class="t-avatar t-avatar--circle"
        >
          <span
            style="transform: scale(1);"
          >
            W
          </span>
        </div>
        <div
          class="t-avatar t-avatar--circle t-avatar-icon"
        >
          <svg
            class="t-icon t-icon-user"
            fill="none"
            height="1em"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M11.5 5a3.5 3.5 0 11-7 0 3.5 3.5 0 017 0zm-1 0a2.5 2.5 0 10-5 0 2.5 2.5 0 005 0zM13.96 10.85c.34.16.54.5.54.87V14a.5.5 0 01-.5.5H2a.5.5 0 01-.5-.5v-2.28c0-.37.2-.7.54-.87a13.79 13.79 0 0111.92 0zM8 10.5c-1.97 0-3.83.45-5.5 1.24v1.76h11v-1.76A12.78 12.78 0 008 10.5z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="demo-avatar-block"
    >
      <div
        class="t-avatar-group t-avatar--offset-right"
      >
        <div
          class="t-avatar t-avatar--circle"
        >
          <img
            src="https://tdesign.gtimg.com/site/avatar.jpg"
          />
        </div>
        <div
          class="t-avatar t-avatar--circle"
        >
          <span
            style="transform: scale(1);"
          >
            W
          </span>
        </div>
        <div
          class="t-avatar t-avatar--circle t-avatar-icon"
        >
          <svg
            class="t-icon t-icon-user"
            fill="none"
            height="1em"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M11.5 5a3.5 3.5 0 11-7 0 3.5 3.5 0 017 0zm-1 0a2.5 2.5 0 10-5 0 2.5 2.5 0 005 0zM13.96 10.85c.34.16.54.5.54.87V14a.5.5 0 01-.5.5H2a.5.5 0 01-.5-.5v-2.28c0-.37.2-.7.54-.87a13.79 13.79 0 0111.92 0zM8 10.5c-1.97 0-3.83.45-5.5 1.24v1.76h11v-1.76A12.78 12.78 0 008 10.5z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`group-cascading.jsx 1`] = `
<DocumentFragment>
  <div
    class="demo-avatar"
  >
    <div
      class="demo-avatar-block"
      style="margin-bottom: 40px;"
    >
      <div
        class="t-avatar-group t-avatar--offset-right"
      >
        <div
          class="t-avatar t-avatar--circle"
        >
          <img
            src="https://tdesign.gtimg.com/site/avatar.jpg"
          />
        </div>
        <div
          class="t-avatar t-avatar--circle"
        >
          <span
            style="transform: scale(1);"
          >
            W
          </span>
        </div>
        <div
          class="t-avatar t-avatar--circle t-avatar-icon"
        >
          <svg
            class="t-icon t-icon-user"
            fill="none"
            height="1em"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M11.5 5a3.5 3.5 0 11-7 0 3.5 3.5 0 017 0zm-1 0a2.5 2.5 0 10-5 0 2.5 2.5 0 005 0zM13.96 10.85c.34.16.54.5.54.87V14a.5.5 0 01-.5.5H2a.5.5 0 01-.5-.5v-2.28c0-.37.2-.7.54-.87a13.79 13.79 0 0111.92 0zM8 10.5c-1.97 0-3.83.45-5.5 1.24v1.76h11v-1.76A12.78 12.78 0 008 10.5z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="demo-avatar-block"
      style="margin-bottom: 40px;"
    >
      <div
        class="t-avatar-group t-avatar--offset-left"
      >
        <div
          class="t-avatar t-avatar--circle"
        >
          <img
            src="https://tdesign.gtimg.com/site/avatar.jpg"
          />
        </div>
        <div
          class="t-avatar t-avatar--circle"
        >
          <span
            style="transform: scale(1);"
          >
            W
          </span>
        </div>
        <div
          class="t-avatar t-avatar--circle t-avatar-icon"
        >
          <svg
            class="t-icon t-icon-user"
            fill="none"
            height="1em"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M11.5 5a3.5 3.5 0 11-7 0 3.5 3.5 0 017 0zm-1 0a2.5 2.5 0 10-5 0 2.5 2.5 0 005 0zM13.96 10.85c.34.16.54.5.54.87V14a.5.5 0 01-.5.5H2a.5.5 0 01-.5-.5v-2.28c0-.37.2-.7.54-.87a13.79 13.79 0 0111.92 0zM8 10.5c-1.97 0-3.83.45-5.5 1.24v1.76h11v-1.76A12.78 12.78 0 008 10.5z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`group-max.jsx 1`] = `
<DocumentFragment>
  <div
    class="demo-avatar"
  >
    <div
      class="demo-avatar-block"
      style="margin-bottom: 40px;"
    >
      <div
        class="t-avatar-group t-avatar--offset-right"
      >
        <div
          class="t-avatar t-avatar--circle"
        >
          <img
            src="https://tdesign.gtimg.com/site/avatar.jpg"
          />
        </div>
        <div
          class="t-avatar t-avatar--circle"
        >
          <span
            style="transform: scale(1);"
          >
            Avatar
          </span>
        </div>
        <div
          class="t-popup__reference"
        >
          <div
            class="t-avatar t-size-l t-avatar--circle"
          >
            <span
              style="transform: scale(1);"
            >
              +1
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="demo-avatar-block"
      style="margin-bottom: 40px;"
    >
      <div
        class="t-avatar-group t-avatar--offset-right"
      >
        <div
          class="t-avatar t-avatar--circle"
        >
          <img
            src="https://tdesign.gtimg.com/site/avatar.jpg"
          />
        </div>
        <div
          class="t-avatar t-avatar--circle"
        >
          <span
            style="transform: scale(1);"
          >
            Avatar
          </span>
        </div>
        <div
          class="t-popup__reference"
        >
          <div
            class="t-avatar t-size-l t-avatar--circle"
          >
            <span
              style="transform: scale(1);"
            >
              <svg
                class="t-icon t-icon-ellipsis"
                fill="none"
                height="1em"
                viewBox="0 0 16 16"
                width="1em"
              >
                <path
                  d="M3 9a1 1 0 110-2 1 1 0 010 2zM7 8a1 1 0 102 0 1 1 0 00-2 0zM12 8a1 1 0 102 0 1 1 0 00-2 0z"
                  fill="currentColor"
                  fill-opacity="0.9"
                />
              </svg>
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="demo-avatar-block"
    >
      <div
        class="t-avatar-group t-avatar--offset-right"
      >
        <div
          class="t-avatar t-avatar--circle"
        >
          <img
            src="https://tdesign.gtimg.com/site/avatar.jpg"
          />
        </div>
        <div
          class="t-avatar t-avatar--circle"
        >
          <span
            style="transform: scale(1);"
          >
            Avatar
          </span>
        </div>
        <div
          class="t-popup__reference"
        >
          <div
            class="t-avatar t-size-l t-avatar--circle"
          >
            <span
              style="transform: scale(1);"
            >
              more
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`shape.jsx 1`] = `
<DocumentFragment>
  <div
    class="demo-avatar"
  >
    <div
      class="t-avatar t-avatar--circle"
      style="margin-right: 40px;"
    >
      <span
        style="transform: scale(1);"
      >
        W
      </span>
    </div>
    <div
      class="t-avatar t-avatar--round"
      style="margin-right: 40px;"
    >
      <span
        style="transform: scale(1);"
      >
        W
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`size.jsx 1`] = `
<DocumentFragment>
  <div
    class="demo-avatar"
  >
    <div
      class="demo-avatar-block"
      style="margin-bottom: 40px;"
    >
      <div
        class="t-avatar t-size-s t-avatar--circle"
        style="margin-right: 40px;"
      >
        <span
          style="transform: scale(1);"
        >
          W
        </span>
      </div>
      <div
        class="t-avatar t-size-m t-avatar--circle"
        style="margin-right: 40px;"
      >
        <span
          style="transform: scale(1);"
        >
          W
        </span>
      </div>
      <div
        class="t-avatar t-size-l t-avatar--circle"
        style="margin-right: 40px;"
      >
        <span
          style="transform: scale(1);"
        >
          W
        </span>
      </div>
      <div
        class="t-avatar t-avatar--circle"
        style="width: 100px; height: 100px; font-size: 50px; margin-right: 40px;"
      >
        <span
          style="transform: scale(1);"
        >
          W
        </span>
      </div>
    </div>
    <div
      class="demo-avatar-block"
    >
      <div
        class="t-avatar t-size-s t-avatar--round"
        style="margin-right: 40px;"
      >
        <span
          style="transform: scale(1);"
        >
          W
        </span>
      </div>
      <div
        class="t-avatar t-size-m t-avatar--round"
        style="margin-right: 40px;"
      >
        <span
          style="transform: scale(1);"
        >
          W
        </span>
      </div>
      <div
        class="t-avatar t-size-l t-avatar--round"
        style="margin-right: 40px;"
      >
        <span
          style="transform: scale(1);"
        >
          W
        </span>
      </div>
      <div
        class="t-avatar t-avatar--round"
        style="width: 100px; height: 100px; font-size: 50px; margin-right: 40px;"
      >
        <span
          style="transform: scale(1);"
        >
          W
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
